app/utilities/schema.py

```diff
@@ -177,20 +191,45 @@ def load_schema_from_url(schema_url, language_code):
 
     constructed_schema_url = f"{schema_url}?language={language_code}"
 
-    req = requests.get(constructed_schema_url)
-    schema_response = req.content.decode()
-    response_duration_in_milliseconds = req.elapsed.total_seconds() * 1000
+    session = requests.Session()
 
-    logger.info(
-        f"schema request took {response_duration_in_milliseconds:.2f} milliseconds",
-        pid=pid,
-    )
+    retries = Retry(
+        total=SCHEMA_REQUEST_MAX_RETRIES,
+        status_forcelist=SCHEMA_REQUEST_RETRY_STATUS_CODES,
+    )  # Codes to retry according to Google Docs https://cloud.google.com/storage/docs/retry-strategy#client-libraries
+
+    retries.BACKOFF_MAX = SCHEMA_REQUEST_MAX_BACKOFF
+
+    session.mount("http://", HTTPAdapter(max_retries=retries))
+    session.mount("https://", HTTPAdapter(max_retries=retries))
+
+    try:
+        req = session.get(constructed_schema_url, timeout=SCHEMA_REQUEST_TIMEOUT)
+    except RequestException as exc:
+        logger.exception(
+            "schema request errored",
+            schema_url=constructed_schema_url,
+        )
+        raise SchemaRequestFailed("schema request failed") from exc
+
+    if req.status_code == 200:
+        schema_response = req.content.decode()
+        response_duration_in_milliseconds = req.elapsed.total_seconds() * 1000
 
-    if req.status_code == 404:
-        logger.error("no schema exists", schema_url=constructed_schema_url)
-        raise NotFound
+        logger.info(
+            f"schema request took {response_duration_in_milliseconds:.2f} milliseconds",
+            pid=pid,
+        )
+
+        return QuestionnaireSchema(json_loads(schema_response), language_code)
+
+    logger.error(
+        "got a non-200 response for schema url request",
+        status_code=req.status_code,
+        schema_url=constructed_schema_url,
+    )
 
-    return QuestionnaireSchema(json_loads(schema_response), language_code)
+    raise SchemaRequestFailed("failed to load schema from url")
```

Comment:
Worth overriding `__str__` of this class and returning `"schema request failed"`, then it can just be `raise SchemaRequestFailed` and message would be consistent. i.e we would like to alert on this in future

Reply:
Done, thanks!